x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/UserRoleMapper.java

```diff
@@ -63,12 +60,8 @@ public UserData(String username, @Nullable String dn, Collection<String> groups,
                         Map<String, Object> metadata, RealmConfig realm) {
             this.username = username;
             this.dn = dn;
-            // noinspection Java9CollectionFactory (because null values happen in some tests, is this realistic?)
-            this.groups = groups == null || groups.isEmpty()
-                    ? Collections.emptySet() : Collections.unmodifiableSet(new HashSet<>(groups));
-            // noinspection Java9CollectionFactory (because null values happen in production code, can such keys be dropped?)
-            this.metadata = metadata == null || metadata.isEmpty()
-                    ? Collections.emptyMap() : Collections.unmodifiableMap(new HashMap<>(metadata));
+            this.groups = Set.copyOf(groups);
+            this.metadata = metadata;
```

Comment:
I don't think it is appropriate for `UserData` to just assume that every caller is passing in an immutable collection.
I believe we should either be copying or wrapping (and I don't think you can safely copy because we might have null values).

Reply:
> I don't think it is appropriate for UserData to just assume that every caller is passing in an immutable collection.

Since the type system is not guaranteeing it, yes it is not safe. In the worst case we can copy on every function.
I will copy it...

> I believe we should either be copying or wrapping (and I don't think you can safely copy because we might have null values).

I believe this is what this PR aims, to assure there are no `null` values?